src/eth-signing/sign-off-chain-action.ts

```diff
@@ -130,11 +141,44 @@ export abstract class SignOffChainAction<M extends {}> extends Signer {
     expectedSigner: Address,
     message: M,
   ): boolean {
-    const hash = this.getHash(message);
-    const signer = ecRecoverTypedSignature(hash, typedSignature);
+    if (stripHexPrefix(typedSignature).length !== 66 * 2) {
+      throw new Error(`Unable to verify signature with invalid length: ${typedSignature}`);
+    }
+
+    const sigType = parseInt(typedSignature.slice(-2), 16);
+    let hashOrMessage: string;
+    switch (sigType) {
+      case SignatureTypes.NO_PREPEND:
+      case SignatureTypes.DECIMAL:
+      case SignatureTypes.HEXADECIMAL:
+        hashOrMessage = this.getHash(message);
+        break;
+      case SignatureTypes.PERSONAL:
+        hashOrMessage = this.getPersonalSignMessage(message);
+        break;
+      default:
+        throw new Error(`Invalid signature type: ${sigType}`);
+    }
+
+    const signer = ecRecoverTypedSignature(hashOrMessage, typedSignature);
     return addressesAreEqual(signer, expectedSigner);
   }
 
+  /**
+   * Get the message string to be signed when using SignatureTypes.PERSONAL.
+   *
+   * This signing method may be used in cases where EIP-712 signing is not possible.
+   * Note: We rely on this function producing a deterministic output for a given input.
+   */
+  public getPersonalSignMessage(
+    message: M,
+  ): string {
+    return JSON.stringify({
+      ...this.getDomainData(),
+      ...message,
+    }, null, 2);
```

Comment:
Slightly worried about the determinism here. Yes, JSON stringify is deterministic, but changing `getDomainData()` or the structure of `message` could ruin the determinism. These inputs could be changed to be "equivalent" objects in a future PR without knowing that it would eventually break something in this function

Reply:
Added much better handling of this, and more test coverage